src/core/cuda_init.cpp

```diff
@@ -49,49 +51,50 @@ struct CompareDevices {
 };
 
 /** Gather list of CUDA devices on all nodes on the master node.
- *  It relies on MPI_Get_processor_name() to get a unique identifier of
- *  the physical node, as opposed to the logical rank of which there can
- *  be more than one on one node.
+ *  It relies on <tt>MPI_Get_processor_name()</tt> to get a unique identifier
+ *  of the physical node, as opposed to the logical rank of which there can
+ *  be more than one per node.
  */
-std::vector<EspressoGpuDevice> cuda_gather_gpus() {
-  int n_gpus = cuda_get_n_gpus();
-  char proc_name[MPI_MAX_PROCESSOR_NAME];
-  int proc_name_len;
+static std::vector<EspressoGpuDevice> mpi_cuda_gather_gpus_local() {
```

Comment:
why do we collect cuda devices if we only support devices on the head node?

Reply:
That's how it was originally written. Should I remove the MPI functionality?

Reply:
no, i was just wondering, why we have the functionality...

Reply:
This functionality was introduced in the Tcl days and wasn't ported to the python interface until now. I wouldn't mind removing the MPI functionality, since we only support devices on the head nodes. It's also cumbersome to find the local MPI rank name in python to get the correct dict key...